### ISSUE TITLE
Build failure with packaged Qt and Wayland integrations

### ISSUE THREAD

**Issue by xvitaly** [telegramdesktop/tdesktop#17349]
### Steps to reproduce

Build with packaged Qt and Wayland integrations.

### Expected behaviour

Successful build.

### Actual behaviour

Build failure:

```
In file included from /builddir/build/BUILD/tdesktop-3.2.5-full/Telegram/lib_waylandshells/waylandshells/xdg-shell/main.cpp:40:
/builddir/build/BUILD/tdesktop-3.2.5-full/Telegram/lib_waylandshells/waylandshells/xdg-shell/qwaylandxdgshellintegration_p.h:70:10: error: 'void WaylandShells::QWaylandXdgShellIntegration::handleKeyboardFocusChanged(QtWaylandClient::QWaylandWindow*, QtWaylandClient::QWaylandWindow*)' marked 'override', but does not override
   70 |     void handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus) override;
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

### Operating system

Fedora 35

### Version of Telegram Desktop

3.2.5 and latest master

### Installation source

Other (unofficial) source

**Comment by john-preston** [comment]
Can you provide the Qt headers version that was used for that build?

**Comment by xvitaly** [comment]
> Can you provide the Qt headers version that was used for that build?

[qt5-qtbase-5.15.2-28.fc35](https://koji.fedoraproject.org/koji/buildinfo?buildID=1854602) - Qt 5.15.2 with KDE patches.

Headers packages:
- https://kojipkgs.fedoraproject.org/packages/qt5-qtbase/5.15.2/28.fc35/x86_64/qt5-qtbase-devel-5.15.2-28.fc35.x86_64.rpm
- https://kojipkgs.fedoraproject.org/packages/qt5-qtbase/5.15.2/28.fc35/x86_64/qt5-qtbase-private-devel-5.15.2-28.fc35.x86_64.rpm

**Comment by ilya-fedin** [comment]
> void handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus) override;

handleKeyboardFocusChanged is present since [Qt 5.8.0](https://github.com/qt/qtwayland/commit/70740da8978cb87b269fcaa73d71595c3c396545). It seems something is wrong with your Qt.

**Comment by xvitaly** [comment]
> handleKeyboardFocusChanged is present since Qt 5.8.0. It seems something is wrong with your Qt.

I can see only one `handleKeyboardFocusChanged()` method in `qwaylanddisplay_p.h`:

```
void handleKeyboardFocusChanged(QWaylandInputDevice *inputDevice);
```

It has another signature.

**Comment by xvitaly** [comment]
https://code.qt.io/cgit/qt/qtwayland.git/tree/src/client/qwaylanddisplay.cpp?id=00390ccf893aa02c8f51e0887624455c7e8d111d#n419

**Comment by ilya-fedin** [comment]
> I can see only one `handleKeyboardFocusChanged()` method in `qwaylanddisplay_p.h`:

Have you looked at the link I provided?

**Comment by xvitaly** [comment]
> Have you looked at the link I provided?

This method is a part of unexported internal implementation, I think.

Header from packaged version of Qt 5.15.2: https://paste.centos.org/view/1dfdd46b

**Comment by mikhailnov** [comment]
```
[user@rosa2019 tmp]$ git clone -b rawhide https://src.fedoraproject.org/rpms/qt5-qtwayland.git
Клонирование в «qt5-qtwayland»…
remote: Enumerating objects: 564, done.
remote: Counting objects: 100% (564/564), done.
remote: Compressing objects: 100% (553/553), done.
remote: Total 564 (delta 204), reused 3 (delta 0), pack-reused 0
Получение объектов: 100% (564/564), 161.74 КиБ | 606.00 КиБ/с, готово.
Определение изменений: 100% (204/204), готово.
[user@rosa2019 tmp]$ cd qt5-qtwayland
[user@rosa2019 qt5-qtwayland]$ grep -inHr handleKeyboardFocusChanged
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:34:@@ -575,14 +575,10 @@ void QWaylandDisplay::handleKeyboardFocusChanged(QWaylandInputDevice *inputDevic
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:39:-        mWaylandIntegration->mShellIntegration->handleKeyboardFocusChanged(keyboardFocus, mLastKeyboardFocus);
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:120:-    virtual void handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus) {
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:137:-void QWaylandXdgShellV5Integration::handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus) {
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:155:-    void handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus) override;
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:167:-void QWaylandXdgShellV6Integration::handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus)
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:192:-    void handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus) override;
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:243:-void QWaylandXdgShellIntegration::handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus)
qtwayland-client-use-wl-keyboard-to-determine-active-state.patch:268:-    void handleKeyboardFocusChanged(QWaylandWindow *newFocus, QWaylandWindow *oldFocus) override;
qtwayland-client-do-not-empty-clipboard-when-new-popup-or-window-is-opened.patch:56:     mParent->mQDisplay->handleKeyboardFocusChanged(mParent);
[user@rosa2019 qt5-qtwayland]$ 

```

**Comment by xvitaly** [comment]
> qtwayland-client-use-wl-keyboard-to-determine-active-state.patch

Thanks. Will ask Fedora maintainers to drop this patch. Closing this.

**Comment by ilya-fedin** [comment]
Well, upstream Qt guarantees QPA to be stable in patch releases and the backport is from another major version. If they patch Qt, I have no other suggestion than patch tdesktop as well

**Comment by xvitaly** [comment]
Upstream MR: https://codereview.qt-project.org/c/qt/qtwayland/+/367306

**Comment by ilya-fedin** [comment]
@Aokromes close this please, this is a downstream bug

**Comment by mikhailnov** [comment]
> Got an answer from Fedora maintainer:
> 
> > That's unfortunate, but this is an upstream change, backported to a stable branch and changes private API, it's not our fault that Telegram is broken

_Many_ Qt5 applications use private API, maintainers of Qt5 would better avoid such games. Here is how many apps use Qt5 private APIs in rosa2021.1, you can count the same for Fedora:
```
# dnf rq --qf '%{sourcerpm}' --whatrequires '*Qt_5.*_PRIVATE_API*' | sort -u | wc -l
68
```

**Comment by xvitaly** [comment]
> close this please, this is a downstream bug

I don't think so. It can be built against 5.15.2, but will fail on 5.15.3+. This is definitely tdesktop issue.

**Comment by ilya-fedin** [comment]
there's no 5.15.3 release, 5.15.2 is the latest public version of Qt 5.15

**Comment by grulja** [comment]
This is not a downstream bug.

This change has been picked and backported to KDE/Qt stable repository (see #1), which means every distribution picks this change automatically if they use this repo for Qt 5.15 backports and many of them do that, including KDE flatpak runtime. 

1) https://invent.kde.org/qt/qt/qtwayland/-/commit/5c180bdc1042e7cb1555e188051f09e219b00ec9

**Comment by ilya-fedin** [comment]
> This change has been picked and backported to KDE/Qt stable repository (see #1)

I see the KDE's version doesn't remove handleKeyboardFocusChanged

**Comment by mikhailnov** [comment]
@xvitaly by the way, this patch https://src.fedoraproject.org/rpms/qt5-qtbase/blob/f35/f/tell-the-truth-about-private-api.patch makes RPM automatically add dependencies from a specific version of qt:
$ rpm -q --requires telegram-desktop | grep PRIVATE
libQt5Core.so.5(Qt_5.15.2_PRIVATE_API)(64bit)
libQt5Gui.so.5(Qt_5.15.2_PRIVATE_API)(64bit)
libQt5WaylandClient.so.5(Qt_5.15.2_PRIVATE_API)(64bit)

I saw that you forced a specific version of Qt5 in Requires in the spec if I remember correctly

**Comment by grulja** [comment]
> > This change has been picked and backported to KDE/Qt stable repository (see #1)
> 
> I see the KDE's version doesn't remove handleKeyboardFocusChanged

It does.

**Comment by grulja** [comment]
Right, I was looking into a wrong comparison. In that case it will get fixed with the next update I'm working on as I dropped the upstream MR version and included the one from KDE/Qt repo.

**Comment by xvitaly** [comment]
Fixed in qt5-qtwayland-devel-5.15.2-15.fc35.
